digester.go

```diff
@@ -83,17 +88,122 @@ func (d *sqlDigester) normalize(sql string) {
 		if pos.Offset == len(sql) {
 			break
 		}
-		if isHead {
-			isHead = false
-		} else {
-			d.buffer.WriteRune(' ')
+		currTok := token{tok, strings.ToLower(lit)}
+		if tok == hintEnd {
+			d.popUntilHintBegin()
+			continue
 		}
-		switch tok {
-		case intLit, stringLit, decLit, floatLit, bitLit, hexLit:
-			d.buffer.WriteRune('?')
-		default:
-			d.buffer.WriteString(strings.ToLower(lit))
+		if isLit(tok) {
+			currTok = d.genericLit(currTok)
 		}
+		d.tokens = append(d.tokens, currTok)
 	}
 	d.lexer.reset("")
+	for i, token := range d.tokens {
+		d.buffer.WriteString(token.lit)
+		if i != len(d.tokens)-1 {
+			d.buffer.WriteRune(' ')
+		}
+	}
+	d.tokens = d.tokens[:0]
+}
+
+func (d *sqlDigester) genericLit(currTok token) token {
+	// "?, ?, ?, ?" => "..."
+	last2 := d.tokens.back(2)
```

Comment:
This means `(1,2,3,x,5,6,7)` and `(1,2,3,4,x,6,7)` will both become `(...,x,...)`?